File: bpf/mount/mount.go
package mount

import (
	"bytes"
	"encoding/binary"
	"github.com/cilium/ebpf/link"
	"github.com/cilium/ebpf/ringbuf"
	"github.com/cilium/ebpf/rlimit"
	"log"
	"sync"
)

// remove -type event if you won't use diy struct in kernel
//
//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -target bpfel -type mount_event bpf mount.c -- -I /sys/kernel/btf
type MountReq struct {
	rd *ringbuf.Reader
}
type MountRes struct {
	Pid     uint32
	DevName [64]byte
	DirName [64]byte
	Type    [32]byte
}

func Start(req *MountReq) (chan MountRes, func()) {
	stopper := make(chan struct{})
	// Allow the current process to lock memory for eBPF resources.
	if err := rlimit.RemoveMemlock(); err != nil {
		log.Fatal(err)
	}
	// Load pre-compiled programs and maps into the kernel.
	objs := bpfObjects{}
	if err := loadBpfObjects(&objs, nil); err != nil {
		log.Fatalf("loading objects: %v", err)
	}
	// write your link code here
	enterLink, err := link.Tracepoint("syscalls", "sys_enter_mount", objs.TraceEnterMount, nil)
	if err != nil {
		log.Fatalf("Failed to attach sys_enter_mount tracepoint: %v", err)
	}
	// 打开 perf buffer
	rd, err := ringbuf.NewReader(objs.Events)
	if err != nil {
		log.Fatalf("Failed to open perf buffer: %v", err)
	}
	req.rd = rd
	out := Action(objs, req, stopper)
	// using closure to create only once close function
	buildClose := func() func() {
		once := sync.Once{}
		return func() {
			once.Do(func() {
				objs.Close()
				enterLink.Close()
				rd.Close()
				// close attach
				close(out)
				close(stopper)
			})
		}
	}

	return out, buildClose()

}
func Action(objs bpfObjects, req *MountReq, stopper chan struct{}) chan MountRes {
	// add your link logic here
	out := make(chan MountRes)
	go func() {
		var event MountRes
		for {
			// write your logical code here
			select {
			case <-stopper:
				return
			default:
				record, err := req.rd.Read()
				if err != nil {
					log.Printf("reading ringbuf: %s", err)
					return
				}
				if err = binary.Read(bytes.NewReader(record.RawSample), binary.LittleEndian, &event); err != nil {
					return
				}
				out <- event
			}
		}
	}()
	return out
}

// add more action function here

File: bpf/mount/mount.c
  //go:build ignore

#include "../vmlinux.h"
#include <bpf/bpf_helpers.h>
#include <bpf/bpf_core_read.h>

char LICENSE[] SEC("license") = "Dual BSD/GPL";

// 定义输出到 perf buffer 的数据结构
struct mount_event {
    __u32 pid;
    char dev_name[64];
    char dir_name[64];
    char type[32];
};
struct mount_event *unused __attribute__((unused));

// 定义 perf buffer
struct {
	__uint(type, BPF_MAP_TYPE_RINGBUF);
	__uint(max_entries, 64 * 1024);
} events SEC(".maps");

// 捕获 sys_enter_mount 事件
SEC("tracepoint/syscalls/sys_enter_mount")
int trace_enter_mount(struct trace_event_raw_sys_enter *ctx) {
    struct mount_event *evt;
    __u32 pid = bpf_get_current_pid_tgid() >> 32;
    evt = bpf_ringbuf_reserve(&events, sizeof(*evt), 0);
    if (!evt){
        return 0;
    }
    evt->pid = pid;
    // 读取挂载参数
    bpf_probe_read_str(&evt->dev_name, sizeof(evt->dev_name), (const char*)ctx->args[0]);
    bpf_probe_read_str(&evt->dir_name, sizeof(evt->dir_name),  (const char*)ctx->args[1]);
    bpf_probe_read_str(&evt->type, sizeof(evt->type),  (const char*)ctx->args[2]);
    // 输出事件到用户空间
    bpf_ringbuf_submit(evt, 0);
    return 0;
}